cloudinit/sources/DataSourceEc2.py

```diff
@@ -947,6 +948,85 @@ def _get_key_as_int_or(dikt, key, alt_value):
     }
 
 
+def _configure_policy_routing(
+    dev_config: dict,
+    *,
+    nic_name: str,
+    nic_metadata: dict,
+    distro: Distro,
+    is_ipv4: bool,
+    table: int,
+) -> None:
+    """
+    Configure policy-based routing on secondary NICs / secondary IPs to
+    ensure outgoing packets are routed via the correct interface.
+
+    @param: dev_config: network cfg v2 to be updated inplace.
+    @param: nic_name: nic name. Only used if ipv4.
+    @param: nic_metadata: nic metadata from IMDS.
+    @param: distro: Instance of Distro. Only used if ipv4.
+    @param: is_ipv4: Boolean indicating if we are acting over ipv4 or not.
+    @param: table: Routing table id.
+    """
+    if not dev_config.get("routes"):
+        dev_config["routes"] = []
+    if is_ipv4:
+        try:
+            lease = distro.dhcp_client.dhcp_discovery(nic_name, distro=distro)
+            gateway = lease["routers"]
+        except NoDHCPLeaseError as e:
+            LOG.warning(
+                "Could not perform dhcp discovery on %s to find its "
+                "gateway. Not adding default route via the gateway. "
+                "Error: %s",
+                nic_name,
+                e,
+            )
+        else:
+            # Add default route via the NIC's gateway
+            dev_config["routes"].append(
+                {
+                    "to": "0.0.0.0/0",
+                    "via": gateway,
+                    "table": table,
+                },
+            )
+
+    if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
```

Comment:
Should this just be consolidated into the above is_ipv4 scope?
```suggestion
            )
        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
```

or Maybe just this diff?
```diff
--- a/cloudinit/sources/DataSourceEc2.py
+++ b/cloudinit/sources/DataSourceEc2.py
@@ -971,6 +971,7 @@ def _configure_policy_routing(
     if not dev_config.get("routes"):
         dev_config["routes"] = []
     if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
         try:
             lease = distro.dhcp_client.dhcp_discovery(nic_name, distro=distro)
             gateway = lease["routers"]
@@ -991,9 +992,6 @@ def _configure_policy_routing(
                     "table": table,
                 },
             )
-
-    if is_ipv4:
-        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
     else:
         subnet_prefix_routes = nic_metadata["subnet-ipv6-cidr-blocks"]
     subnet_prefix_routes = (
```